File: src/NdtBundleService/Services/ZplNdtLabelBuilder.cs
using System.Text;
using NdtBundleService.Configuration;

namespace NdtBundleService.Services;

/// <summary>
/// Builds ZPL for the full NDT bundle tag (Honeywell PD45S).
/// Layout matches the physical NDT tag:
/// - Top: Code 128 barcode with NDT Batch Number (human-readable line printed by the barcode command)
/// - Middle content (4–5 lines):
///   Mill, PO Number, NDT Batch Number on one line
///   Grade, Pipe Size, Pipe Thickness on one line
///   Pipe Length and bundle total weight (kg) on one line
///   Date, Number of NDT pipes, Pipe type/WIP/FG and optional "Reprint" on one line
/// - Bottom: one or two stacked Code 128 barcodes with the same NDT Batch Number.
/// </summary>
public static class ZplNdtLabelBuilder
{
    /// <summary>203 dpi Honeywell PD45S: ~8 dots/mm.</summary>
    public const int DotsPerMm = 8;

    public readonly record struct NdtTagLabelSize(int WidthMm, int LengthMm)
    {
        public int WidthDots => Math.Max(DotsPerMm, WidthMm * DotsPerMm);
        public int LengthDots => Math.Max(DotsPerMm, LengthMm * DotsPerMm);

        public bool IsSquare => LengthMm >= 100;

        public static NdtTagLabelSize FromOptions(NdtBundleOptions options, int millNo = 0)
        {
            var widthMm = options.NdtTagLabelWidthMm;
            var lengthMm = options.NdtTagLabelLengthMm;
            if (millNo >= 1
                && millNo <= 4
                && options.NdtTagLabelLengthMmByMill.TryGetValue(millNo.ToString(), out var overrideLength)
                && overrideLength > 0)
            {
                lengthMm = overrideLength;
            }

            return new NdtTagLabelSize(widthMm, lengthMm);
        }
    }

    public static byte[] BuildNdtTagZpl(
        string ndtBatchNo,
        int millNo,
        string poNumber,
        string? pipeGrade,
        string pipeSize,
        string pipeThickness,
        string pipeLength,
        string bundleWeight,
        string pipeType,
        DateTime date,
        int pcsInBundle,
        bool isReprint,
        string? stationText = null,
        NdtTagLabelSize? labelSize = null)
    {
        var size = labelSize ?? NdtTagLabelSize.FromOptions(new NdtBundleOptions());
        return size.IsSquare
            ? BuildSquareTagZpl(
                ndtBatchNo,
                millNo,
                poNumber,
                pipeGrade,
                pipeSize,
                pipeThickness,
                pipeLength,
                bundleWeight,
                pipeType,
                date,
                pcsInBundle,
                isReprint,
                stationText,
                size)
            : BuildCompactTagZpl(
                ndtBatchNo,
                millNo,
                poNumber,
                pipeGrade,
                pipeSize,
                pipeThickness,
                pipeLength,
                bundleWeight,
                pipeType,
                date,
                pcsInBundle,
                isReprint,
                stationText,
                size);
    }

    private static byte[] BuildCompactTagZpl(
        string ndtBatchNo,
        int millNo,
        string poNumber,
        string? pipeGrade,
        string pipeSize,
        string pipeThickness,
        string pipeLength,
        string bundleWeight,
        string pipeType,
        DateTime date,
        int pcsInBundle,
        bool isReprint,
        string? stationText,
        NdtTagLabelSize size)
    {
        var zpl = new StringBuilder();
        zpl.Append("^XA");
        zpl.AppendFormat("^PW{0}^LL{1}^LH0,0", size.WidthDots, size.LengthDots);

        var escapedBatch = Escape(ndtBatchNo);
        var escapedPo = Escape(poNumber);
        var escapedGrade = Escape(pipeGrade);
        var escapedSize = Escape(pipeSize);
        var escapedThickness = Escape(pipeThickness);
        var escapedLength = Escape(pipeLength);
        var escapedWeight = Escape(bundleWeight);
        var escapedType = Escape(pipeType);
        var escapedStation = Escape(stationText);

        const int left = 60;
        var fieldWidth = size.WidthDots - (left * 2);

        var y = 12;
        zpl.AppendFormat("^FO{0},{1}^BY2^BCN,55,Y,N,N^FD{2}^FS", left, y, escapedBatch);
        y += 72;

        zpl.Append("^CF0,22");
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDMill- {3}  PO: {4}  Bund: {5}^FS",
            left, y, fieldWidth, millNo, escapedPo, escapedBatch);
        y += 24;

        var gradePart = string.IsNullOrEmpty(escapedGrade) ? "Gr- -" : $"Gr- {escapedGrade}";
        var sizePart = string.IsNullOrEmpty(escapedSize) ? "-" : escapedSize;
        var thkPart = string.IsNullOrEmpty(escapedThickness) ? "-" : escapedThickness;
        var lenPart = string.IsNullOrEmpty(escapedLength) ? "-" : escapedLength;
        var wtPart = string.IsNullOrEmpty(escapedWeight) ? "-" : escapedWeight;

        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FD{3}  Size: {4}  Thk: {5}^FS",
            left, y, fieldWidth, gradePart, sizePart, thkPart);
        y += 24;
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDLen: {3}  Wt: {4}^FS", left, y, fieldWidth, lenPart, wtPart);
        y += 24;

        if (!string.IsNullOrWhiteSpace(escapedStation))
        {
            zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDStation: {3}^FS", left, y, fieldWidth, escapedStation);
            y += 24;
        }

        var dateText = date.ToString("dd/MM/yy");
        var typeText = string.IsNullOrEmpty(escapedType) ? "" : $"  {escapedType}";
        var reprintText = isReprint ? "  Reprint" : "";
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDDate: {3}  Pcs. {4}{5}{6}^FS",
            left, y, fieldWidth, dateText, pcsInBundle, typeText, reprintText);

        var bottomY = size.LengthDots - 78;
        zpl.AppendFormat("^FO{0},{1}^BY2^BCN,55,Y,N,N^FD{2}^FS", left, bottomY, escapedBatch);

        zpl.Append("^XZ");
        return Encoding.UTF8.GetBytes(zpl.ToString());
    }

    private static byte[] BuildSquareTagZpl(
        string ndtBatchNo,
        int millNo,
        string poNumber,
        string? pipeGrade,
        string pipeSize,
        string pipeThickness,
        string pipeLength,
        string bundleWeight,
        string pipeType,
        DateTime date,
        int pcsInBundle,
        bool isReprint,
        string? stationText,
        NdtTagLabelSize size)
    {
        var zpl = new StringBuilder();
        zpl.Append("^XA");
        zpl.AppendFormat("^PW{0}^LL{1}^LH0,0", size.WidthDots, size.LengthDots);
        zpl.Append("^CF0,34");

        var escapedBatch = Escape(ndtBatchNo);
        var escapedPo = Escape(poNumber);
        var escapedGrade = Escape(pipeGrade);
        var escapedSize = Escape(pipeSize);
        var escapedThickness = Escape(pipeThickness);
        var escapedLength = Escape(pipeLength);
        var escapedWeight = Escape(bundleWeight);
        var escapedType = Escape(pipeType);
        var escapedStation = Escape(stationText);

        const int left = 60;
        var fieldWidth = size.WidthDots - (left * 2);

        var y = 36;
        zpl.AppendFormat("^FO{0},{1}^BY3^BCN,110,Y,N,N^FD{2}^FS", left, y, escapedBatch);
        y += 148;

        zpl.Append("^CF0,30");
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDMill- {3}  PO: {4}  Bund: {5}^FS",
            left, y, fieldWidth, millNo, escapedPo, escapedBatch);
        y += 38;

        var gradePart = string.IsNullOrEmpty(escapedGrade) ? "Gr- -" : $"Gr- {escapedGrade}";
        var sizePart = string.IsNullOrEmpty(escapedSize) ? "-" : escapedSize;
        var thkPart = string.IsNullOrEmpty(escapedThickness) ? "-" : escapedThickness;
        var lenPart = string.IsNullOrEmpty(escapedLength) ? "-" : escapedLength;
        var wtPart = string.IsNullOrEmpty(escapedWeight) ? "-" : escapedWeight;

        zpl.Append("^CF0,28");
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FD{3}  Size: {4}  Thk: {5}^FS",
            left, y, fieldWidth, gradePart, sizePart, thkPart);
        y += 36;
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDLen: {3}  Wt: {4}^FS", left, y, fieldWidth, lenPart, wtPart);
        y += 36;

        if (!string.IsNullOrWhiteSpace(escapedStation))
        {
            zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDStation: {3}^FS", left, y, fieldWidth, escapedStation);
            y += 36;
        }

        zpl.Append("^CF0,32");
        var dateText = date.ToString("dd/MM/yy");
        var typeText = string.IsNullOrEmpty(escapedType) ? "" : $"  {escapedType}";
        var reprintText = isReprint ? "  Reprint" : "";
        zpl.AppendFormat("^FO{0},{1}^FB{2},1,0,C,0^FDDate: {3}  Pcs. {4}{5}{6}^FS",
            left, y, fieldWidth, dateText, pcsInBundle, typeText, reprintText);

        var bottomY1 = size.LengthDots - 250;
        var bottomY2 = bottomY1 + 118;
        zpl.AppendFormat("^FO{0},{1}^BY2^BCN,88,Y,N,N^FD{2}^FS", left, bottomY1, escapedBatch);
        zpl.AppendFormat("^FO{0},{1}^BY2^BCN,88,Y,N,N^FD{2}^FS", left, bottomY2, escapedBatch);

        zpl.Append("^XZ");
        return Encoding.UTF8.GetBytes(zpl.ToString());
    }

    private static string Escape(string? value)
    {
        if (string.IsNullOrEmpty(value)) return "";
        return value
            .Replace("\\", "\\\\")
            .Replace("^", "\\^")
            .Replace("~", "\\~");
    }
}

File: plc-server/plc/zplNdtLabelBuilder.js
"use strict";

const DotsPerMm = 8;

function escape(value) {
  if (value == null || value === "") return "";
  return String(value)
    .replace(/\\/g, "\\\\")
    .replace(/\^/g, "\\^")
    .replace(/~/g, "\\~");
}

function resolveLabelSize(widthMm = 100, lengthMm = 100) {
  return {
    widthDots: Math.max(DotsPerMm, widthMm * DotsPerMm),
    lengthDots: Math.max(DotsPerMm, lengthMm * DotsPerMm),
    isSquare: lengthMm >= 100,
  };
}

function buildCompactTagZpl(params, size) {
  const {
    ndtBatchNo,
    millNo,
    poNumber,
    pipeGrade,
    pipeSize,
    pipeThickness,
    pipeLength,
    pipeWeightPerMeter,
    pipeType,
    date,
    pcsInBundle,
    isReprint,
    stationText,
  } = params;

  const zpl = [];
  zpl.push("^XA");
  zpl.push(`^PW${size.widthDots}^LL${size.lengthDots}^LH0,0`);

  const escapedBatch = escape(ndtBatchNo);
  const escapedPo = escape(poNumber);
  const escapedGrade = escape(pipeGrade);
  const escapedSize = escape(pipeSize);
  const escapedThickness = escape(pipeThickness);
  const escapedLength = escape(pipeLength);
  const escapedWeight = escape(pipeWeightPerMeter);
  const escapedType = escape(pipeType);
  const escapedStation = escape(stationText);

  const left = 60;
  const fieldWidth = size.widthDots - left * 2;

  let y = 12;
  zpl.push(`^FO${left},${y}^BY2^BCN,55,Y,N,N^FD${escapedBatch}^FS`);
  y += 72;

  zpl.push("^CF0,22");
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDMill- ${millNo}  PO: ${escapedPo}  Bund: ${escapedBatch}^FS`
  );
  y += 24;

  const gradePart = escapedGrade === "" ? "Gr- -" : `Gr- ${escapedGrade}`;
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FD${gradePart}  Size: ${
      escapedSize === "" ? "-" : escapedSize
    }  Thk: ${escapedThickness === "" ? "-" : escapedThickness}^FS`
  );
  y += 24;
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDLen: ${
      escapedLength === "" ? "-" : escapedLength
    }  Wt: ${escapedWeight === "" ? "-" : escapedWeight}^FS`
  );
  y += 24;

  if (escapedStation) {
    zpl.push(`^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDStation: ${escapedStation}^FS`);
    y += 24;
  }

  const d = date instanceof Date ? date : new Date();
  const dd = String(d.getDate()).padStart(2, "0");
  const mm = String(d.getMonth() + 1).padStart(2, "0");
  const yy = String(d.getFullYear()).slice(-2);
  const dateText = `${dd}/${mm}/${yy}`;
  const typeText = escapedType ? `  ${escapedType}` : "";
  const reprintText = isReprint ? "  Reprint" : "";
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDDate: ${dateText}  Pcs. ${pcsInBundle}${typeText}${reprintText}^FS`
  );

  const bottomY = size.lengthDots - 78;
  zpl.push(`^FO${left},${bottomY}^BY2^BCN,55,Y,N,N^FD${escapedBatch}^FS`);
  zpl.push("^XZ");
  return Buffer.from(zpl.join(""), "utf8");
}

function buildSquareTagZpl(params, size) {
  const {
    ndtBatchNo,
    millNo,
    poNumber,
    pipeGrade,
    pipeSize,
    pipeThickness,
    pipeLength,
    pipeWeightPerMeter,
    pipeType,
    date,
    pcsInBundle,
    isReprint,
    stationText,
  } = params;

  const margin = 40;
  const left = margin;
  const fieldWidth = size.widthDots - margin * 2;
  const topBarcodeHeight = 145;
  const bottomBarcodeHeight = 100;
  const barcodeHumanReadable = 34;
  const bottomGap = 16;

  const bottomY2 = size.lengthDots - margin - bottomBarcodeHeight - barcodeHumanReadable;
  const bottomY1 = bottomY2 - bottomGap - bottomBarcodeHeight - barcodeHumanReadable;
  const textStartY = margin + topBarcodeHeight + barcodeHumanReadable + 12;
  const textEndY = bottomY1 - 12;
  const hasStation = Boolean(stationText);
  const textLineCount = hasStation ? 5 : 4;
  const lineStep = Math.max(46, Math.floor((textEndY - textStartY) / textLineCount));

  const zpl = [];
  zpl.push("^XA");
  zpl.push(`^PW${size.widthDots}^LL${size.lengthDots}^LH0,0`);

  const escapedBatch = escape(ndtBatchNo);
  const escapedPo = escape(poNumber);
  const escapedGrade = escape(pipeGrade);
  const escapedSize = escape(pipeSize);
  const escapedThickness = escape(pipeThickness);
  const escapedLength = escape(pipeLength);
  const escapedWeight = escape(pipeWeightPerMeter);
  const escapedType = escape(pipeType);
  const escapedStation = escape(stationText);

  let y = margin;
  zpl.push(`^FO${left},${y}^BY3^BCN,${topBarcodeHeight},Y,N,N^FD${escapedBatch}^FS`);
  y = textStartY;

  zpl.push("^CF0,38");
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDMill- ${millNo}  PO: ${escapedPo}  Bund: ${escapedBatch}^FS`
  );
  y += lineStep;

  const gradePart = escapedGrade === "" ? "Gr- -" : `Gr- ${escapedGrade}`;
  zpl.push("^CF0,34");
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FD${gradePart}  Size: ${
      escapedSize === "" ? "-" : escapedSize
    }  Thk: ${escapedThickness === "" ? "-" : escapedThickness}^FS`
  );
  y += lineStep;
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDLen: ${
      escapedLength === "" ? "-" : escapedLength
    }  Wt: ${escapedWeight === "" ? "-" : escapedWeight}^FS`
  );
  y += lineStep;

  if (hasStation) {
    zpl.push(`^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDStation: ${escapedStation}^FS`);
    y += lineStep;
  }

  const d = date instanceof Date ? date : new Date();
  const dd = String(d.getDate()).padStart(2, "0");
  const mm = String(d.getMonth() + 1).padStart(2, "0");
  const yy = String(d.getFullYear()).slice(-2);
  const dateText = `${dd}/${mm}/${yy}`;
  const typeText = escapedType ? `  ${escapedType}` : "";
  const reprintText = isReprint ? "  Reprint" : "";
  zpl.push("^CF0,36");
  zpl.push(
    `^FO${left},${y}^FB${fieldWidth},1,0,C,0^FDDate: ${dateText}  Pcs. ${pcsInBundle}${typeText}${reprintText}^FS`
  );

  zpl.push(`^FO${left},${bottomY1}^BY3^BCN,${bottomBarcodeHeight},Y,N,N^FD${escapedBatch}^FS`);
  zpl.push(`^FO${left},${bottomY2}^BY3^BCN,${bottomBarcodeHeight},Y,N,N^FD${escapedBatch}^FS`);
  zpl.push("^XZ");
  return Buffer.from(zpl.join(""), "utf8");
}

/**
 * Port of ZplNdtLabelBuilder.BuildNdtTagZpl (UTF-8 bytes for Honeywell / raw TCP).
 */
function buildNdtTagZpl({
  ndtBatchNo,
  millNo,
  poNumber,
  pipeGrade,
  pipeSize,
  pipeThickness,
  pipeLength,
  pipeWeightPerMeter,
  pipeType,
  date,
  pcsInBundle,
  isReprint,
  stationText,
  labelWidthMm = 100,
  labelLengthMm = 100,
}) {
  const size = resolveLabelSize(labelWidthMm, labelLengthMm);
  const params = {
    ndtBatchNo,
    millNo,
    poNumber,
    pipeGrade,
    pipeSize,
    pipeThickness,
    pipeLength,
    pipeWeightPerMeter,
    pipeType,
    date,
    pcsInBundle,
    isReprint,
    stationText,
  };

  return size.isSquare ? buildSquareTagZpl(params, size) : buildCompactTagZpl(params, size);
}

module.exports = { buildNdtTagZpl, resolveLabelSize, DotsPerMm };
